### PR TITLE
Use `bitflags` to represent allowed `FeatureBlocks`

### PR DESCRIPTION
**Pull request by grtlr**
# Description of change

Changes `ALLOWED_FEATURE_BLOCKS` to a `bitflags`-based representation and adapts the errors for a malformed list of `FeatureBlocks`.  

## Type of change

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Change checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes
- [x] I have updated the CHANGELOG.md, if my changes are significant enough

### CODE REVIEW

File: bee-message/src/output/feature_block/mod.rs [hunk 1]
```diff
@@ -293,25 +308,50 @@ fn validate_dependencies(feature_blocks: &[FeatureBlock]) -> Result<(), Error> {
 
 pub(crate) fn validate_allowed_feature_blocks(
     feature_blocks: &FeatureBlocks,
-    allowed_feature_blocks_kind: &[u8],
+    allowed_feature_blocks: FeatureBlockUsages,
 ) -> Result<(), Error> {
-    if feature_blocks.len() > allowed_feature_blocks_kind.len() {
-        return Err(Error::TooManyFeatureBlocks {
```
Comment (thibault-martinez) [inline bee-message/src/output/feature_block/mod.rs:299]: Why was this removed ?

File: bee-message/src/output/feature_block/mod.rs [hunk 2]
```diff
@@ -293,25 +308,50 @@ fn validate_dependencies(feature_blocks: &[FeatureBlock]) -> Result<(), Error> {
...
-
-    for feature_block in feature_blocks.iter() {
-        index = allowed_feature_blocks_kind[index..]
-            .binary_search(&feature_block.kind())
-            .map_err(|index| Error::UnallowedFeatureBlock {
+    let mut blocks = FeatureBlockUsages::empty();
+    for (index, feature_block) in feature_blocks.iter().enumerate() {
+        let usage = feature_block.usage();
+        if blocks.contains(usage) {
+            return Err(Error::DuplicateFeatureBlock {
```
Comment (thibault-martinez) [inline bee-message/src/output/feature_block/mod.rs]: You don't need to check for duplicates, an already constructed `FeatureBlocks` ensures that there are no duplicates.

File: bee-message/src/output/alias.rs [hunk 3]
```diff
@@ -157,12 +154,14 @@ impl AliasOutput {
...
 
-    ///
-    const ALLOWED_FEATURE_BLOCKS: [u8; 3] = [
-        SenderFeatureBlock::KIND,
-        IssuerFeatureBlock::KIND,
-        MetadataFeatureBlock::KIND,
-    ];
+    /// Returns the set of allowed [`FeatureBlock`]s for an [`AliasOutput`].
+    #[inline(always)]
+    pub fn allowed_feature_blocks() -> FeatureBlockUsages {
```
Comment (thibault-martinez) [inline bee-message/src/output/alias.rs]: Some of your `allowed_feature_blocks` methods are public and some are not. I don't think this is something that needs to be public, I'd remove it unless proven useful. Also, these can be const methods.

File: bee-message/src/output/extended.rs [hunk 4]
```diff
@@ -4,18 +4,13 @@
...
-            SenderFeatureBlock, TimelockMilestoneIndexFeatureBlock, TimelockUnixFeatureBlock,
-        },
+        feature_block::{validate_allowed_feature_blocks, FeatureBlock, FeatureBlockUsages, FeatureBlocks},
         NativeToken, NativeTokens,
     },
     Error,
 };
 
 use bee_common::packable::{Packable, Read, Write};
-
```
Comment (thibault-martinez) [inline bee-message/src/output/extended.rs]: Any reason to remove this ?

File: bee-message/src/output/extended.rs [hunk 5]
```diff
@@ -95,17 +90,20 @@ pub struct ExtendedOutput {
...
-        ExpirationMilestoneIndexFeatureBlock::KIND,
-        ExpirationUnixFeatureBlock::KIND,
-        MetadataFeatureBlock::KIND,
-        IndexationFeatureBlock::KIND,
-    ];
+
+    /// Returns the set of allowed [`FeatureBlock`]s for an [`ExtendedOutput`].
+    #[inline(always)]
+    pub fn allowed_feature_blocks() -> FeatureBlockUsages {
+        FeatureBlockUsages::empty()
```
Comment (thibault-martinez) [inline bee-message/src/output/extended.rs]: Is the empty necessary ?

File: bee-message/src/output/feature_block/mod.rs [hunk 6]
```diff
@@ -293,25 +308,50 @@ fn validate_dependencies(feature_blocks: &[FeatureBlock]) -> Result<(), Error> {
...
+        /// Signals the presence of a [`ExpirationMilestoneIndexFeatureBlock`].
+        const EXPIRATION_MILESTONE_INDEX = 1 << ExpirationMilestoneIndexFeatureBlock::KIND;
+        /// Signals the presence of a [`ExpirationUnixFeatureBlock`].
+        const EXPIRATION_UNIX = 1 << ExpirationUnixFeatureBlock::KIND;
+        /// Signals the presence of a [`MetadataFeatureBlock`].
+        const METADATA = 1 << MetadataFeatureBlock::KIND;
+        /// Signals the presence of a [`IndexationFeatureBlock`].
+        const INDEXATION = 1 << IndexationFeatureBlock::KIND;
+        /// Signals the presence of a [`IssuerFeatureBlock`].
+        const ISSUER = 1 << IssuerFeatureBlock::KIND;
```
Comment (thibault-martinez) [inline bee-message/src/output/feature_block/mod.rs]: Can you put them in their actual order please ? I think issuer is after sender

File: bee-message/src/output/alias.rs (hunk 3 above)
Comment (grtlr) [reply bee-message/src/output/alias.rs]: I tried making them `const` before, but `bitflags` creates a trait and trait methods can't be marked `const` yet. I'll change the visibility though.

File: bee-message/src/output/extended.rs (hunk 5 above)
Comment (grtlr) [reply bee-message/src/output/extended.rs]: It's not necessary, but I like it for readability reasons. I can remove it though, your call.

File: bee-message/src/output/feature_block/mod.rs (hunk 1 above)
Comment (grtlr) [reply bee-message/src/output/feature_block/mod.rs:299]: I thought we could never run into this case if feature blocks are sorted, unique, and allowed?

File: bee-message/src/output/feature_block/mod.rs (hunk 2 above)
Comment (grtlr) [reply bee-message/src/output/feature_block/mod.rs]: Even better 😄

File: bee-message/src/output/feature_block/mod.rs [hunk 7]
```diff
@@ -293,25 +308,41 @@ fn validate_dependencies(feature_blocks: &[FeatureBlock]) -> Result<(), Error> {
...
                 kind: feature_block.kind(),
-            })?;
+            });
+        }
     }
-
     Ok(())
 }
+
+use bitflags::bitflags;
```
Comment (thibault-martinez) [inline bee-message/src/output/feature_block/mod.rs]: Can you put this with the other imports ?

File: bee-message/src/output/feature_block/mod.rs [hunk 8]
```diff
@@ -73,6 +73,21 @@ impl FeatureBlock {
             Self::Metadata(_) => MetadataFeatureBlock::KIND,
         }
     }
+
+    /// Returns the [`FeatureBlockUsages`] for the given [`FeatureBlock`].
```
Comment (thibault-martinez) [inline bee-message/src/output/feature_block/mod.rs]: No longer called FeatureBlockUsages

Comment (thibault-martinez) [reply bee-message/src/output/feature_block/mod.rs]: Why did this pass the docs CI ?

File: bee-message/src/output/feature_block/mod.rs [hunk 9]
```diff
@@ -73,6 +73,21 @@ impl FeatureBlock {
...
+    pub(crate) fn usage(&self) -> FeatureBlockFlags {
```
Comment (thibault-martinez) [inline bee-message/src/output/feature_block/mod.rs]: I think the name is a tiny bit weird, can we think of something else ?

Comment (thibault-martinez) [reply bee-message/src/output/feature_block/mod.rs]: Simply `flag` ?

File: bee-message/src/output/feature_block/mod.rs (hunk 8 above)
Comment (grtlr) [reply bee-message/src/output/feature_block/mod.rs]: Hmm, that's weird indeed. I'll look into it.

File: bee-test/src/rand/output/feature_block.rs [hunk 10]
```diff
@@ -0,0 +1,65 @@
+// Copyright 2020-2021 IOTA Stiftung
```
Comment (thibault-martinez) [inline bee-test/src/rand/output/feature_block.rs]: I assume this wasn't meant to be added 😬  ?

**Review by thibault-martinez — APPROVED** [review]
